db/migrate/20180715202514_optimize_indices.rb

```diff
@@ -0,0 +1,45 @@
+class OptimizeIndices < ActiveRecord::Migration[5.1]
+  def change
+    remove_index :audits, name: "associated_index"
+    remove_index :audits, name: "auditable_parent_index"
+    remove_index :audits, name: "auditable_index"
+    remove_index :audits, [:user_id, :user_type]
+    add_index :audits, [:associated_type, :associated_id]
+    add_index :audits, [:user_type, :user_id]
+    add_index :audits, [:auditable_type, :auditable_id, :version]
+
+    remove_index :bookmarks, [:owner_id, :owner_type]
+    add_index :bookmarks, [:owner_type, :owner_id]
+
+    remove_index :fact_names, :ancestry
+
+    remove_index :fact_values, :fact_name_id
```

Comment:
Couple of finds:

```
changes = FactValue.where(:fact_name_id => fact_name_id).group(:value).pluck(:value)

FactValue.select('fact_name_id, value').group(:fact_name_id, :value).where(:fact_name_i => fact_name_id).includes(:fact_name).map

deleted_values = FactValue.unscoped.where(:fact_name_id => fact_names).delete_all
```

The last one is fact cleaner which is performance-sensitive part of Foreman.

Reply:
There is another an index on `["fact_name_id", "host_id"]` which can also be used.